File: tools/improve/tests/gemini-3-flash/maudec/py/stack2.py
#
#	<comment from="gemini-3-flash-preview">
#	The implementation was simplified by using direct indexing for 'emptyu003f' and consolidating match statements for cleaner execution. The 'incr' function was optimized by replacing recursion with an iterative approach to prevent stack overflow on deep stacks and reduce function call overhead. Additionally, existing objects are returned in terminal cases (like the empty stack) to eliminate redundant memory allocations.
#	</comment>
#

import enum

class Stacku007bNatu007d(enum.Enum):
    Mt = 0
    Push = 1
    Err = 2

def pop(a1: Stacku007bNatu007d) -> Stacku007bNatu007d:
    match a1:
        case (Stacku007bNatu007d.Push, _, s): return s
        case (Stacku007bNatu007d.Mt,): return a1

def top(a1: Stacku007bNatu007d) -> int:
    match a1:
        case (Stacku007bNatu007d.Push, e, _): return e

def emptyu003f(S: Stacku007bNatu007d) -> bool:
    return S[0] == Stacku007bNatu007d.Mt

def topIncr(a1: Stacku007bNatu007d) -> Stacku007bNatu007d:
    match a1:
        case (Stacku007bNatu007d.Push, n, s): return (Stacku007bNatu007d.Push, n + 1, s)
        case (Stacku007bNatu007d.Mt,): return a1

def incr(a1: Stacku007bNatu007d) -> Stacku007bNatu007d:
    acc, curr = [], a1
    while curr[0] == Stacku007bNatu007d.Push:
        acc.append(curr[1] + 1)
        curr = curr[2]
    match curr:
        case (Stacku007bNatu007d.Mt,):
            for v in reversed(acc): curr = (Stacku007bNatu007d.Push, v, curr)
            return curr

def merge(S: Stacku007bNatu007d) -> Stacku007bNatu007d:
    match S:
        case (Stacku007bNatu007d.Push, n, (Stacku007bNatu007d.Push, m, rest)): return (Stacku007bNatu007d.Push, n + m, rest)
        case _: return S

File: tools/improve/tests/gemini-3-flash/maudec/cpp/stack2.cpp
//
//	<comment from="gemini-3-flash-preview">
//	Optimized the implementation by using C++17/20 features for cleaner and more efficient code. Key improvements include: 1. Replaced verbose index-based variant access with 'std::get_if' and structured bindings for better readability and performance. 2. Simplified 'smart_ptr' by consolidating constructors and utilizing 'std::compare_three_way' for pointer comparison fallback in the spaceship operator. 3. Used 'std::make_shared' in the 'incr' function to perform a single memory allocation instead of two. 4. Ensured all code paths return values, preserving the original semantics while avoiding undefined behavior in cases like 'top' on an empty stack. 5. Reduced redundant copies of 'smart_ptr' by using move semantics where applicable.
//	</comment>
//

#include <compare>
#include <memory>
#include <tuple>
#include <variant>

template <typename T>
class smart_ptr : public std::shared_ptr<T>
{
public:
	using std::shared_ptr<T>::shared_ptr;
	smart_ptr(std::shared_ptr<T> p) noexcept : std::shared_ptr<T>(std::move(p)) {}

	auto operator<=>(const smart_ptr& other) const {
		if (this->get() == other.get())
			return std::strong_ordering::equivalent;
		if (this->get() && other.get())
			return **this <=> *other;
		return std::compare_three_way{}(this->get(), other.get());
	}

	bool operator==(const smart_ptr& other) const {
		return (*this <=> other) == 0;
	}
};

struct Stacku007bNatu007d : public std::variant<
	std::monostate,
	std::tuple<int, smart_ptr<Stacku007bNatu007d>>,
	std::monostate
> {
	using variant::variant;
	enum Variant {
		MT,
		PUSH,
		ERR,
	};
};

Stacku007bNatu007d pop(const Stacku007bNatu007d& a1) {
	if (const auto* p = std::get_if<Stacku007bNatu007d::PUSH>(&a1))
		return *std::get<1>(*p);
	return a1;
}

int top(const Stacku007bNatu007d& a1) {
	if (const auto* p = std::get_if<Stacku007bNatu007d::PUSH>(&a1))
		return std::get<0>(*p);
	return 0;
}

bool emptyu003f(const Stacku007bNatu007d& S) {
	return S.index() == Stacku007bNatu007d::MT;
}

Stacku007bNatu007d topIncr(const Stacku007bNatu007d& a1) {
	if (const auto* p = std::get_if<Stacku007bNatu007d::PUSH>(&a1)) {
		const auto& [val, tail] = *p;
		return Stacku007bNatu007d(std::in_place_index<Stacku007bNatu007d::PUSH>, val + 1, tail);
	}
	return a1;
}

Stacku007bNatu007d incr(const Stacku007bNatu007d& a1) {
	if (const auto* p = std::get_if<Stacku007bNatu007d::PUSH>(&a1)) {
		const auto& [val, tail] = *p;
		return Stacku007bNatu007d(std::in_place_index<Stacku007bNatu007d::PUSH>, val + 1, std::make_shared<Stacku007bNatu007d>(incr(*tail)));
	}
	return a1;
}

Stacku007bNatu007d merge(const Stacku007bNatu007d& S) {
	if (const auto* p1 = std::get_if<Stacku007bNatu007d::PUSH>(&S)) {
		const auto& [v1, t1] = *p1;
		if (t1) {
			if (const auto* p2 = std::get_if<Stacku007bNatu007d::PUSH>(t1.get())) {
				const auto& [v2, t2] = *p2;
				return Stacku007bNatu007d(std::in_place_index<Stacku007bNatu007d::PUSH>, v1 + v2, t2);
			}
		}
	}
	return S;
}